### PR TITLE
Added code for Insertion in m-ary trees

### PR DESCRIPTION
**Pull request by Saptashrungi**

### CODE REVIEW

File: pydatastructs/trees/m_ary_trees.py [hunk 1]
```diff
@@ -77,7 +77,13 @@ def insert(self, key, data):
...
         None
         """
-        raise NotImplementedError("This is an abstract method.")
+        temp = self.search(key)
+        if temp.is_root is True and temp.key is None:
+            temp.key = key
+            temp.data = data
+        if temp is not None:
+            self.tree[temp].data = data
+
```
Comment (czgdp1807) [inline pydatastructs/trees/m_ary_trees.py]: Please see https://github.com/codezonediitj/pydatastructs/issues/101#issuecomment-595095244

File: pydatastructs/utils/misc_util.py [hunk 2]
```diff
@@ -297,3 +298,26 @@ def __new__(cls, key, data):
...
+    __slots__ = ['keys', 'degree', 'children', 'data', 'is_root', 'no_of_keys', 'search', 'traverse']
+
+    def __new__(cls, key, data, degree, children):
+        obj = Node.__new__(cls)
+        obj.data, obj.key = data, key
+        obj.degree = degree
+        obj.no_of_keys = no_of_keys
+        obj.children = [] or children
+        obj.is_root = False
+        return obj
```
Comment (czgdp1807) [inline pydatastructs/utils/misc_util.py]: https://github.com/codezonediitj/pydatastructs/blob/692624605aeb8a47c33f82e4838f8a01b8af5d3d/pydatastructs/utils/misc_util.py#L136-L143

No of keys can be figured out from no of children.
Is degree the same as order in Knuth's definition.

File: pydatastructs/trees/m_ary_trees.py (hunk 1 above)
Comment (Saptashrungi) [reply pydatastructs/trees/m_ary_trees.py]: So we need only search function for m-ary?

**Comment by czgdp1807** [comment]
@vibhu18116 is already working on `MArySearchTree`. In addition, APIs for B-Tree depend on `MArySearchTree`. So, let's wait for @vibhu18116 to complete `MArySearchTree` first. 
Closing.

File: pydatastructs/trees/m_ary_trees.py (hunk 1 above)
Comment (czgdp1807) [reply pydatastructs/trees/m_ary_trees.py]: `MAryTree` is abstract. `MArySearchTree` is concrete in the way insertion and search is done, so both of these method will be defined in that class.

File: pydatastructs/utils/misc_util.py (hunk 2 above)
Comment (Saptashrungi) [reply pydatastructs/utils/misc_util.py]: Degree represents the lower bound on the number of children a node in the B Tree can have (except for the root). i.e the minimum number of children possible. Whereas the Order represents the upper bound on the number of children. ie. the maximum number possible.

Comment (czgdp1807) [reply pydatastructs/utils/misc_util.py]: > Every non-leaf node (except root) has at least ⌈m/2⌉ child nodes.

Can degree be computed from order?

**Pull request closed without merging**